iml-agent/src/action_plugins/check_ha.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{agent_error::ImlAgentError, cmd::cmd_output_success};
+use elementtree::Element;
+use futures::{future, Future};
+//use tracing_futures::Instrument;
+use std::collections::HashMap;
+
+// standard:provider:ocftype (e.g. ocf:heartbeat:ZFS, or stonith:fence_ipmilan)
+#[derive(serde::Deserialize, serde::Serialize, Clone, Debug)]
+pub struct ResourceAgentType {
+    pub standard: String,         // e.g. ocf, lsb, stonith, etc..
+    pub provider: Option<String>, // e.g. heartbeat, lustre, chroma
+    pub ocftype: String,          // e.g. Lustre, ZFS
+}
+
+impl ResourceAgentType {
+    pub fn new(standard: String, provider: Option<String>, ocftype: String) -> Self {
+        ResourceAgentType {
+            standard,
+            provider,
+            ocftype,
+        }
+    }
+}
+
+impl From<ResourceAgentType> for String {
+    fn from(agent: ResourceAgentType) -> Self {
+        match agent.provider {
+            Some(provider) => format!("{}:{}:{}", agent.standard, provider, agent.ocftype),
+            None => format!("{}:{}", agent.standard, agent.ocftype),
+        }
+    }
+}
+
+#[derive(serde::Deserialize, serde::Serialize, Clone, Debug)]
+pub struct AgentInfo {
+    pub agent: ResourceAgentType,
+    pub group: Option<String>,
+    pub id: String,
+    pub args: HashMap<String, String>,
+}
+
+impl AgentInfo {
+    pub fn new(elem: &Element, group: Option<String>) -> AgentInfo {
+        AgentInfo {
+            agent: ResourceAgentType::new(
+                elem.get_attr("class").unwrap().to_string(),
```

Comment:
Should we replace these `unwraps` with a conversion to a specific error type? If so, then we can use a type similar to:

https://github.com/whamcloud/integrated-manager-for-lustre/blob/0aa188ef4c4d954784a9144c6b4923f3cdbfa751/iml-agent/src/agent_error.rs#L10-L11

Where we can put the name of the attribute that was missing in the field.

That way, we avoid panicing and can return the missing field name to the caller.

Reply:
I could not coerce this into working correctly.  Trying to use "?" from within the `flat_map()` in `check_ha()` would throw an error:
```
error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> iml-agent/src/action_plugins/check_ha.rs:88:67
   |
88 |                         .chain(elem.find_all("primitive").map(|p| AgentInfo::create(p, None)?))
   |                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `action_plugins::check_ha::AgentInfo`
   |
   = help: the trait `std::ops::Try` is not implemented for `action_plugins::check_ha::AgentInfo`
   = note: required by `std::ops::Try::from_error`
```